### ISSUE TITLE
Differentiation for Concatenation of different types

### ISSUE THREAD

**Issue by thobe** [opencypher/openCypher#189]
In implementations of Cypher today (Neo4j), the `+` operator is overloaded with multiple meanings:

* `+` is used to mean numeric addition.
* `+` is used to mean concatenations of strings.
  - In the case of concatenation of strings, there is also an implicit conversion of non-string values to string values in order to achieve string concatenation.
* `+` is used to mean concatenation of lists: i.e. appending one list with another.
  - In the case of list concatenation, there is also an implicit conversion of non-list values into a list containing a single element, that element being the other operand of the `+` operator.

There is no problem with each of these individual meanings on its own. There also isn't much problem with having these three different usages of `+`, at least as long as only the base meaning is being considered, except of course that it is not obvious from reading a query whether a `+` operator is used to mean numeric addition, string concatenation, or list concatenation.
The real problem with this overloaded behaviour comes from the implicit coercion rules, and the fact that they are actually in conflict with one another.

Considering the two expressions `someString + someList` and `someList + someString` (where `someString` is a string and `someList` is a list) it is not clear if `someString` should be wrapped in a list in order to perform list concatenation, or if `someList` should be converted to a string in order to perform string concatenation. In fact the Neo4j implementation even performs different conversions in different scenarios for the very same expression. Even if the precedence between these coercions was more explicit, there is not necessarily a way to explicitly express one interpretation or the other.

We need a clear specification for the semantics of these different operations.

**Comment by thobe** [comment]
I see three possible resolutions for this:

1. Keep the three different meanings of `+`, and the coercion behaviour of them, but clearly specify the precedence between the coercion rules.
2. Keep the three different meanings of `+`, but require the operands to both be of compatible types, and never perform implicit coercion.
3. Introduce new operators for appending strings, and for appending lists, so that these three different operations use three different operators.

**Comment by boggle** [comment]
This has the potential for a lot of breakage, my spontaneous reaction is:

(1) list + string means add the string as an element to the list
(2) if you want to use the list as a string, use a function for that (toString? str?), i.e. remove the coercion behavior

Regading (2), the question is how much coercion we want to retain. I'd be ok if we still allowed number + string but let's not extend this to containers.

**Comment by Mats-SX** [comment]
I'm interested in exploring a variation of (3): add new operator for list concatenation (`++`), and keep `+` to mean both string concatenation and numeric addition. The rules for `+` could be:

- Only scalar operands are valid. (`toString()` supports all scalars.)
- If and only if both operands are numeric, the operation is numeric addition.
- Otherwise, at least one operand must be a string.
- If the other operand is not a string, a `toString()` call will implicitly be inserted around the other operand (coercion).

For `++`:

- At least one operand must be a list.
- If the other operand is not a list, the operation is list extension (prepend or append, as determined by operand order).

In general, I feel like (3) (or this variation of it) comes with the best advantage, in that it aids readability (and thus understandability) of queries.

**Comment by boggle** [comment]
@Mats-SX I think long-term this is great, I'm somewhat concerned about the amount of breakage we will introduce by removing all existing overloads at once. Perhaps the right idea is to combine (2), (3), i.e. select a sensible, self-consistent set of overloads for `+` based on what is more frequently used and introduce `++`. Then start generating warnings favoring `++` over `+`.

**Comment by Mats-SX** [comment]
That sounds like a transition/implementation concern to me. For vendors who do not have many existing users with queries that rely on the current behaviour, that will not be an issue.

**Comment by petraselmer** [comment]
I agree with (3) being the cleanest and most future-proof option going forwards, following the rules @Mats-SX suggests
